server/src/main/java/org/opensearch/index/store/IndexStoreListener.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.store;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.opensearch.common.annotation.PublicApi;
+import org.opensearch.core.index.Index;
+import org.opensearch.core.index.shard.ShardId;
+import org.opensearch.env.NodeEnvironment;
+import org.opensearch.index.IndexSettings;
+
+import java.util.List;
+
+/**
+ * A listener that is executed on per-index and per-shard store events, like deleting shard path
+ *
+ * @opensearch.api
+ */
+@PublicApi(since = "2.19.0")
+public interface IndexStoreListener {
+    default void beforeShardPathDeleted(ShardId shardId, IndexSettings indexSettings, NodeEnvironment env) {}
+
+    default void beforeIndexPathDeleted(Index index, IndexSettings indexSettings, NodeEnvironment env) {}
+
+    IndexStoreListener EMPTY = new IndexStoreListener() {
+    };
+
+    /**
+     * A Composite listener that multiplexes calls to each of the listeners methods.
+     */
+    final class CompositeIndexStoreListener implements IndexStoreListener {
```

Comment:
Any reasons to have this class public?

Reply:
For this I was referencing `SearchOperationListener` which defines the inner class in the same way:
https://github.com/opensearch-project/OpenSearch/blob/978d14e867c41b224303b2e77745495336568f79/server/src/main/java/org/opensearch/index/shard/SearchOperationListener.java#L180-L185

I think it makes sense to keep it in the interface especially as it's in a different package from where it's used in `Node.java`

Reply:
> I think it makes sense to keep it in the interface especially as it's in a different package from where it's used in Node.java

We could make it package private, or do we expect it to be used by plugin developers?

Reply:
I don't expect a plugin developer to use the composite listener, but since we create the composite listener in `Node.java` if it's package private then it also needs to be in the `package org.opensearch.node` which doesn't seem right to me.

Reply:
Make it inner class instead? Since it is contained to `Node` only?

Reply:
That would work, but that was my original point about how all the other listener type interfaces today just have a public inner composite class:
* https://github.com/opensearch-project/OpenSearch/blob/3d1fa9848f5a83e689c0eed1039613125d4f8fdf/server/src/main/java/org/opensearch/action/search/SearchRequestOperationsListener.java#L74
* https://github.com/opensearch-project/OpenSearch/blob/09bacee5fc85676e97bee6b4ad87dec35c6aa8cc/server/src/main/java/org/opensearch/index/shard/IndexingOperationListener.java#L97
* https://github.com/opensearch-project/OpenSearch/blob/978d14e867c41b224303b2e77745495336568f79/server/src/main/java/org/opensearch/index/shard/SearchOperationListener.java#L183

Also thinking about it some more I think it's good to expose the composite listener to plugins since if a single plugin wants to implement multiple `IndexStoreListener`s they can do so via a composite listener if we leave it public -- I'm guessing this is why the other listener types do this as well.

Reply:
Sure, thanks @jed326